google/cloud/storage/internal/common_metadata.h

```diff
@@ -72,8 +73,11 @@ class CommonMetadata {
  public:
   CommonMetadata() : metageneration_(0), owner_() {}
 
-  static CommonMetadata<Derived> ParseFromJson(internal::nl::json const& json) {
-    CommonMetadata<Derived> result{};
+  static Status ParseFromJson(CommonMetadata<Derived>& result,
```

Comment:
Why not StatusOr on this one?  Or am I jumping the gun?

Reply:
I think the code gets uglier with a `StatusOr<>`, but maybe I am wrong, I think it would need to be:

```C++
auto status = CommonMetadata<BucketMetadata>::ParseFromJson(json);
if (not status.ok()) {
  return std::move(status).status();
}
BucketMetadata result;
reinterpret_cast<CommonMetadata<BucketMetadata>&>(result) = std::move(*status);
```

as it is, the [usage](https://github.com/GoogleCloudPlatform/google-cloud-cpp/pull/1741/files#diff-e55dee86f957dc1440cf8ca3b2e9b212R120) of this function is a bit simpler to grok:

```C++
   BucketMetadata result{};
   auto status = CommonMetadata<BucketMetadata>::ParseFromJson(result, json);
   if (not status.ok()) {
     return status;
   }
```

Since this is a non-public interface, in the `internal` namespace, and used only in two places I went with the easier code from the calling point. Thoughts?

Reply:
SGTM.  My eyes are too focused on just groking the diffs for now that the public/internal distinction is still blurry.  Thanks.